crates/astria-sequencer/src/app/mod.rs

```diff
@@ -147,37 +147,161 @@ const POST_TRANSACTION_EXECUTION_RESULT_KEY: &str = "post_transaction_execution_
+            next_validators_hash: request.next_validators_hash,
+            misbehavior: request.misbehavior,
+            txs: response.txs,
+        };
+        let pb_data = pb::RequestProcessProposal::from(proposal).encode_to_vec();
+
+        self.data = Some(Sha256::digest(pb_data).into());
+        self.status = ProposalFingerprintStatus::PreparedProposal;
+        Ok(())
+    }
+
+    // Given a ProcessProposal request, check the ProcessProposal matches
+    // the fingerprint from prepare_proposal. If in `PreparedProposal` state
+    // will create a partial ProcessProposal message, serialize and hash that
+    // to compare. If it does not match, the status is set to `InvalidChecked`.
+    // The overall status is returned.
+    fn check_if_prepared_proposal(
+        &mut self,
+        proposal: &abci::request::ProcessProposal,
+    ) -> ProposalFingerprintStatus {
+        match self.status {
+            ProposalFingerprintStatus::Unset
+            | ProposalFingerprintStatus::InvalidChecked
+            | ProposalFingerprintStatus::ExecutedBlock => {}
+            ProposalFingerprintStatus::PreparedProposal => {
+                use sha2::Sha256;
+                use tendermint_proto::v0_38::abci as pb;
+                let partial_proposal = abci::request::ProcessProposal {
+                    hash: Hash::default(),
+                    ..proposal.clone()
+                };
+                let pb_data = pb::RequestProcessProposal::from(partial_proposal).encode_to_vec();
+                if self.data != Some(Sha256::digest(pb_data).into()) {
+                    self.status = ProposalFingerprintStatus::InvalidChecked;
+                }
+            }
+        };
+
+        self.status
+    }
+
+    // Called after process_proposal has been called or finalize_block to set
+    // to a ExecutedBlock fingerprint. Can only be called on a PreparedProposal
+    // or Unset fingerprint, otherwise will error.
+    fn set_executed_block(&mut self, proposal_hash: [u8; 32]) -> Result<()> {
+        match self.status {
+            ProposalFingerprintStatus::Unset | ProposalFingerprintStatus::PreparedProposal => {
+                self.data = Some(proposal_hash);
+                self.status = ProposalFingerprintStatus::ExecutedBlock;
+            }
+            ProposalFingerprintStatus::ExecutedBlock => {
+                bail!("executed block fingerprint shouldn't be set twice");
+            }
+            ProposalFingerprintStatus::InvalidChecked => {
+                bail!("executed block fingerprint shouldn't be set after invalid check");
+            }
         }
+
+        Ok(())
+    }
+
+    // Given a proposal hash, check if the proposal hash matches the fingerprint
+    // If checking against an ExecutedBlock fingerprint, will compare the hash, update
+    // the status to InvalidChecked if it does not match.
+    // Cannot be called on a PreparedProposal fingerprint, will change status
+    // to InvalidChecked.
```

Comment:
this can (and often will) be called on a PreparedProposal fingerprint - say we propose a block, then `process_proposal` is called with that block, then status is `PreparedProposal` and fingerprint status remains unchanged. then when we call `finalize_block` on that block then `InvalidChecked` is returned.

Reply:
nevermind,`set_execute_block` is called in `post_execute_transactions`, which is called in both cases of `process_proposal`. the app is getting a bit hard to reason about